plugins/CoreHome/vue/src/EntityDuplicator/EntityDuplicatorAdapter.ts

```diff
@@ -0,0 +1,164 @@
+
+  /**
+   * Optional: Called after failed duplication
+   */
+  onFailure?(error: DuplicateRequestResponse | Error): void;
+
+  /**
+   * Optional: Called before showing the modal and can be used to show the modal as loading if
+   * there are any async operations needed to finish presenting the modal.
+   */
+  beforeShowModal?(): void | Promise<void>;
+}
+
+export class BaseDuplicatorAdapter implements EntityDuplicatorAdapter {
+  module: string;
+
+  method: string;
+
+  format: string;
+
+  requiredFields: string[];
+
+  constructor(properties: EntityDuplicatorAdapterProperties) {
+    this.module = properties.module || 'API';
+    this.method = properties.method;
+    this.format = properties.format || 'json';
+    this.requiredFields = properties.requiredFields || ['idSite', 'idDestinationSites'];
+  }
+
+  async validateFormFields(
+    formValues: Record<string, unknown>,
+  ): Promise<ValidationResult> {
+    const errorMessages: string[] = [];
+
+    this.requiredFields.forEach((fieldName) => {
+      if (!(fieldName in formValues) || !formValues[fieldName]) {
+        errorMessages.push(translate('General_Required', fieldName));
+      }
+    });
+
+    return new Promise((resolve) => resolve({
+      errorMessages,
+      isValid: errorMessages.length === 0,
+    } as ValidationResult));
+  }
+
+  prepareApiParams(formValues: Record<string, unknown>): QueryParameters {
+    return {
+      idSite: Matomo.idSite || MatomoUrl.parsed.value.idSite,
+      idDestinationSites: [formValues.idDestinationSite as number|string],
+      ...formValues,
+    };
+  }
+
+  async submitRequest(params: QueryParameters): Promise<DuplicateRequestResponse> {
+    // Override the defaults if provided in the params and then remove them from the params
+    this.module = params.module || this.module;
+    this.method = params.method || this.method;
+    this.format = params.format || this.format;
+    const postParams = params as Omit<QueryParameters, 'module' | 'method' | 'format'>;
+
+    if (!this.method || this.method.length < 1) {
+      throw new Error('The POST method cannot be empty!');
+    }
+
+    const ajax = new AjaxHelper();
+    // Remove some default parameters as they aren't applicable to copying existing reports
+    ajax.removeDefaultParameter('date');
+    ajax.removeDefaultParameter('period');
+    ajax.removeDefaultParameter('segment');
+    // Set the main params as part of the URL
+    ajax.addParams({
+      module: this.module,
+      method: this.method,
+      format: this.format,
+    }, 'GET');
+    ajax.addParams(postParams, 'POST');
+    ajax.setFormat(this.format);
+    return ajax.send();
+  }
+
+  /**
+   * Optional: Called during onSuccess method if both are defined. This allows using the default
+   * success method while also defining some custom actions such as reloading a store
+   */
+  onSuccessReloadStore?(response: DuplicateRequestResponse): Promise<void>;
```

Comment:
```suggestion
  onSuccessCallback?(response: DuplicateRequestResponse): Promise<void>;
```

Or something different, but still generic. The documentation itself already hints it can be used for anything, so we should not include a "reload store" intent in the name.

Reply:
Yeah. I debated about renaming it. I'll do that. Thank you for the suggestion.